csrc/cpu/comm/riscv64/shm_rvv.cpp

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0
+
+// DeepSpeed Team
+
+#include <torch/extension.h>
+#include <cmath>
+#include "shm.h"
+
+using float16_t = _Float16;
+using float32_t = float_t;
+
+inline vfloat32m2_t cvt_bf16_to_fp32(vuint16m1_t src, size_t vl)
+{
+    vuint32m2_t widened = __riscv_vwcvtu_x_x_v_u32m2(src, vl);
+    return __riscv_vreinterpret_v_u32m2_f32m2(__riscv_vsll_vx_u32m2(widened, 16, vl));
+}
+
+inline vuint16m1_t cvt_fp32_to_bf16(vfloat32m2_t src, size_t vl)
+{
+    vuint32m2_t value = __riscv_vreinterpret_v_f32m2_u32m2(src);
+    vuint32m2_t nan = __riscv_vmv_v_x_u32m2(0xFFFF, vl);
+    vbool16_t mask_value = __riscv_vmfne_vv_f32m2_b16(src, src, vl);
+    vuint32m2_t ones = __riscv_vmv_v_x_u32m2(0x1, vl);
+    vuint32m2_t vec_bias = __riscv_vmv_v_x_u32m2(0x7FFF, vl);
+    // uint32_t lsb = (input >> 16) & 1;
+    vuint32m2_t t_value = __riscv_vand_vx_u32m2(__riscv_vsrl_vx_u32m2(value, 16, vl), 0x1, vl);
+    // uint32_t rounding_bias = 0x7fff + lsb;
+    t_value = __riscv_vadd_vv_u32m2(t_value, vec_bias, vl);
+    // input += rounding_bias;
+    t_value = __riscv_vadd_vv_u32m2(t_value, value, vl);
+    // input = input >> 16;
+    t_value = __riscv_vsrl_vx_u32m2(t_value, 16, vl);
+    // Check NaN before converting back to bf16
+    t_value = __riscv_vmerge_vvm_u32m2(t_value, nan, mask_value, vl);
+
+    return __riscv_vncvt_x_x_w_u16m1(t_value, vl);
+}
+
+inline vfloat32m2_t cvt_fp16_to_fp32(vfloat16m1_t src, size_t vl)
+{
+    return __riscv_vfwcvt_f_f_v_f32m2(src, vl);
+}
+
+inline vfloat16m1_t cvt_fp32_to_fp16(vfloat32m2_t src, size_t vl)
+{
+    return __riscv_vfncvt_rod_f_f_w_f16m1(src, vl);
+}
+
+#define CVT_ADD_BF16(x)                                                                   \
+    do {                                                                                  \
+        auto in##x##_val =                                                                \
+            cvt_bf16_to_fp32(__riscv_vle16_v_u16m1((uint16_t*)(buffers[x] + i), vl), vl); \
+        inout_val = __riscv_vfadd_vv_f32m2(inout_val, in##x##_val, vl);                   \
+    } while (0)
+
+// Reduce functions down below use vectorized algorithm, the number of bytes processed each
+// iteration depends on vector length. 128bit vector ==> 16 bytes, 256bit vector ==> 32 bytes
+void reduce_bf16_buffers(int start_elements, int num_elements, char* to_buffer, char** buffers)
```

Comment:
I feel that `reduce_xx_buffers` are more than a simple building block needs to be implemented in different CPU instructions set.  It has >50 lines of code and most logic is the same between `riscv64` and `x86_64`.  Is it possible for these functions moved to shared file and call abstract lower level operations (load, store, add) instead?